### ISSUE TITLE
Material Passport: timestamps of events

### ISSUE THREAD

**Issue by sbocconi** [reflow-project/weloop#106]
The Material Passport lacks timestamps of the events, which seems important for a passport if you want to know whether your resources are freshly produced for example, or 10 years ago.

**Comment by pral2a** [comment]
Yes, that's a well known issue. Currently the backend supports `hasBeginning`, `hasEnd` and `hasPointInTime` properties for each `economicEvent` but all of them have to be filled in front-end, the backend only persist the provided `DateTime`  without further validation of the `economicEvent` business logic. 

We' need to discuss this with Dyne, otherwise we'll implement it in front-end only.

See example:

```graphql
{
  economicResource(id: "01FCT51592SSVQG8WC64GM5ZYC") {
    id
    name
    primaryAccountable {
      name
    }
    track {
      hasBeginning
      hasEnd
      hasPointInTime 
      resourceInventoriedAs {
        id
        name
        note
      }
      toResourceInventoriedAs {
        id
        name
        note
      }
      action {
        id
        label
      }
  		provider {
        name
        id
      }
      receiver {
        name
        id
      }
    }
    trace {
      hasBeginning
      hasEnd
      hasPointInTime
      toResourceInventoriedAs {
        id
        name
        note
      }
      action {
        id
        label
      }
      id
      provider {
        name
        id
      }
      receiver {
        name
        id
      }
    }
  }
}
```

Result:

```graphql
{

  "data": {
    "economicResource": {
      "id": "01FCT51592SSVQG8WC64GM5ZYC",
      "name": "Gibson Les Paul Studio",
      "primaryAccountable": {
        "name": "ocataco"
      },
      "trace": [],
      "track": [
        {
          "action": {
            "id": "transfer",
            "label": "transfer"
          },
          "hasBeginning": null,
          "hasEnd": null,
          "hasPointInTime": null,
          "provider": {
            "id": "01F9BG4Y8DE13MWT8RR3ZFWKQ7",
            "name": "ocataco"
          },
          "receiver": {
            "id": "01F5ZVRW4TY3C6M49FYZVHSXZ0",
            "name": "Stefano Bocconi"
          },
          "resourceInventoriedAs": {
            "id": "01FCT51592SSVQG8WC64GM5ZYC",
            "name": "Gibson Les Paul Studio",
            "note": "I think that I should be able to define a provider here, for example a Music Shop, even if the provider does not have an account on WeLoop.\nWhy am I (Stefano Bocconi) not the owner of the resource?\nPlus the Unit are not working, I only see liter and kilo.\nWho is the real owner?"
          },
          "toResourceInventoriedAs": {
            "id": "01FCTNQZQVSHN050609HXA5DJV",
            "name": "Gibson Les Paul Studio",
            "note": "I think that I should be able to define a provider here, for example a Music Shop, even if the provider does not have an account on WeLoop.\nWhy am I (Stefano Bocconi) not the owner of the resource?\nPlus the Unit are not working, I only see liter and kilo."
          }
        },
        {
          "action": {
            "id": "transfer",
            "label": "transfer"
          },
          "hasBeginning": null,
          "hasEnd": null,
          "hasPointInTime": null,
          "provider": {
            "id": "01F9BG4Y8DE13MWT8RR3ZFWKQ7",
            "name": "ocataco"
          },
          "receiver": {
            "id": "01F5ZVRW4TY3C6M49FYZVHSXZ0",
            "name": "Stefano Bocconi"
          },
          "resourceInventoriedAs": {
            "id": "01FCT51592SSVQG8WC64GM5ZYC",
            "name": "Gibson Les Paul Studio",
            "note": "I think that I should be able to define a provider here, for example a Music Shop, even if the provider does not have an account on WeLoop.\nWhy am I (Stefano Bocconi) not the owner of the resource?\nPlus the Unit are not working, I only see liter and kilo.\nWho is the real owner?"
          },
          "toResourceInventoriedAs": null
        },
        {
          "action": {
            "id": "transfer",
            "label": "transfer"
          },
          "hasBeginning": null,
          "hasEnd": null,
          "hasPointInTime": null,
          "provider": {
            "id": "01F9BG4Y8DE13MWT8RR3ZFWKQ7",
            "name": "ocataco"
          },
          "receiver": {
            "id": "01F5ZVRW4TY3C6M49FYZVHSXZ0",
            "name": "Stefano Bocconi"
          },
          "resourceInventoriedAs": {
            "id": "01FCT51592SSVQG8WC64GM5ZYC",
            "name": "Gibson Les Paul Studio",
            "note": "I think that I should be able to define a provider here, for example a Music Shop, even if the provider does not have an account on WeLoop.\nWhy am I (Stefano Bocconi) not the owner of the resource?\nPlus the Unit are not working, I only see liter and kilo.\nWho is the real owner?"
          },
          "toResourceInventoriedAs": {
            "id": "01FCTTR56J12PJGTECQ9VH0PAC",
            "name": "Gibson Les Paul Studio",
            "note": "I think that I should be able to define a provider here, for example a Music Shop, even if the provider does not have an account on WeLoop.\nWhy am I (Stefano Bocconi) not the owner of the resource?\nPlus the Unit are not working, I only see liter and kilo.\nWho is the real owner?"
          }
        }
      ]
    }
  }
}
```

**Comment by sbocconi** [comment]
Sorry @pral2a I am not sure I understand.

Are you saying that the back-end should be filling the `hasPointInTime` property as it is more something the backend should be in charge to determine, to avoid clients messing around with it?

And I cannot find any `DateTime` in the message dump recorded by the back-end, is this something that is shown somewhere else?

**Comment by pral2a** [comment]
> Are you saying that the back-end should be filling the hasPointInTime property as it is more something the backend should be in charge to determine, to avoid clients messing around with it?
 
Yes, exactly, and that is currently not happening. See

```graphql
          "hasBeginning": null,
          "hasEnd": null,
          "hasPointInTime": null,
```

However, as I said we can always fix it temporary by posting the client time in front-end. We asked Dyne and we're waiting to know more about it.

**Comment by sbocconi** [comment]
I see two scenarios where different options could be useful (not necessarily via the Weloop interface but also via the GraphQL one):

- When monitoring and certifying material flows, it sounds as a back-end given timestamp makes more sense
- When importing data (from another system, or simulated) for testing or seeding the system with existing data, it is useful to be able to determine the point in time.

Maybe to support both use cases there should be a client time and a server time, which could be used in different situations.

**Comment by pral2a** [comment]
Yes, we agree both might be necessary. We created an issue in backend to complement this one https://github.com/dyne/zenpub/issues/61

**Comment by pral2a** [comment]
Fixed via #149
